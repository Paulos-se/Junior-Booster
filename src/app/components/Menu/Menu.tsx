import Button from '@mui/material/Button';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import Fade from '@mui/material/Fade';
import { useState } from 'react';
import { Box, Link, ListItemIcon } from '@mui/material';
import * as Icons from '@mui/icons-material';
import { SvgIconComponent } from '@mui/icons-material';

export interface MenuItemProps {
  id: string;
  label: string;
  url: string;
  icon: string;
}
export interface MenuProps {
  /**
   * MenuItems should contain an object: <br/>
   * id: string;<br/>
   * label: string;<br/>
   * url: string;<br/>
   * icon: string; from MUI icons https://mui.com/material-ui/material-icons/;
   */
  menuItems: MenuItemProps[];
  color: string;
}

/** Menu */
/**
 * FadeMenu Component
 *
 * The `FadeMenu` component is a versatile and dynamic menu that uses Material-UI components to provide a smooth fade transition.
 * It accepts an array of menu items, each with an identifier, label, URL, and an icon from the Material-UI icon set.
 * This component is designed to enhance the user interface by offering a dropdown menu with customizable options and icons.
*/
export const FadeMenu = ({ menuItems, color }: MenuProps) => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <>
      {/* \/ replace to new button \/ */}
      <Button
        id="fade-button"
        aria-controls={open ? 'fade-menu' : undefined}
        aria-haspopup="true"
        aria-expanded={open ? 'true' : undefined}
        onClick={handleClick}
      >
        Button
      </Button>
      {/* /\ replace to new button /\ */}

      <Menu
        id="fade-menu"
        MenuListProps={{
          'aria-labelledby': 'fade-button',
        }}
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        TransitionComponent={Fade}
      >
        {menuItems.map((item) => (
          <MenuItem key={item.id} onClick={handleClose}>
            <Link href={item.url} sx={{ textDecoration: 'none' }}>
              <Box sx={{ display: 'flex', color: { color } }}>
                {IconMenu(item.icon)}
                {item.label}
              </Box>
            </Link>
          </MenuItem>
        ))}
      </Menu>
    </>
  );
};

function IconMenu(icon: string) {
  const getIconComponent = (iconName: string): SvgIconComponent | null => {
    const IconComponent = Icons[iconName as keyof typeof Icons];
    if (IconComponent) {
      return IconComponent;
    }
    return null;
  };
  const IconComponent = getIconComponent(icon);

  return (
    <>
      {IconComponent ? (
        <ListItemIcon>
          <IconComponent />
        </ListItemIcon>
      ) : null}
    </>
  );
}
