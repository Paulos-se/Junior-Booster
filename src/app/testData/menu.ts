import { MenuProps } from '../components/Menu/Menu';

export const menuItemsMock: MenuProps = {
  menuItems: [
    { id: '1', label: 'Home', url: '/', icon: 'Home' },
    { id: '2', label: 'About', url: '/about', icon: 'Info' },
    { id: '3', label: 'Contact', url: '/contact', icon: 'ContactMail' },
  ],
  color: 'grey',
};
