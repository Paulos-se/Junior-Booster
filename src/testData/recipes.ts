export const mockRecipes = {
  recipes: [
    {
      id: 652825,
      title: 'My Asian Calamari',
      image: 'https://img.spoonacular.com/recipes/652825-556x370.jpg',
      imageType: 'jpg',
      summary: `My Asian Calamari requires around <b>45 minutes</b> from start to finish. Watching your figure? This dairy free and pescatarian recipe has <b>563 calories</b>, <b>24g of protein</b>, and <b>5g of fat</b> per serving. For <b>$2.6 per serving</b>, this recipe <b>covers 23%</b> of your daily requirements of vitamins and minerals. This recipe serves 4. This recipe is typical of Asian cuisine. A mixture of all spice powder, cornstarch, squid rings, and a handful of other ingredients are all it takes to make this recipe so delicious. 3 people have made this recipe and would make it again. It works well as a rather inexpensive hor d'oeuvre. It is brought to you by Foodista. Taking all factors into account, this recipe <b>earns a spoonacular score of 59%</b>, which is solid. Users who liked this recipe also liked <a href="https://spoonacular.com/recipes/asian-calamari-salad-586727">Asian Calamari Salad</a>, <a href="https://spoonacular.com/recipes/southeast-asian-rice-noodles-with-calamari-and-herbs-188585">Southeast Asian Rice Noodles with Calamari and Herbs</a>, and <a href="https://spoonacular.com/recipes/christmas-eve-calamari-calamari-in-umido-per-la-vigilia-di-natale-597284">Christmas Eve Calamari | Calamari in Umido per la Vigilia di Natale</a>.`,
    },
    {
      id: 632778,
      title: 'Artisan Farfalle Pasta With Smoked Salmon and Cream Sauce',
      image: 'https://img.spoonacular.com/recipes/632778-556x370.jpg',
      summary:
        'Artisan Farfalle Pasta With Smoked Salmon and Cream Sauce might be just the main course you are searching for. This recipe makes 2 servings with <b>762 calories</b>, <b>28g of protein</b>, and <b>48g of fat</b> each. For <b>$6.5 per serving</b>, this recipe <b>covers 31%</b> of your daily requirements of vitamins and minerals. It is brought to you by Foodista. Only a few people made this recipe, and 3 would say it hit the spot. From preparation to the plate, this recipe takes approximately <b>45 minutes</b>. If you have chicken broth, artisan pasta, extra virgin oil, and a few other ingredients on hand, you can make it. Overall, this recipe earns a <b>pretty good spoonacular score of 78%</b>. Similar recipes include <a href="https://spoonacular.com/recipes/farfalle-pasta-with-smoked-salmon-and-cream-cheese-1233493">Farfalle Pasta With Smoked Salmon And Cream Cheese</a>, <a href="https://spoonacular.com/recipes/farfalle-pasta-with-smoked-salmon-and-cream-cheese-30071">Farfalle Pasta With Smoked Salmon And Cream Cheese</a>, and <a href="https://spoonacular.com/recipes/smoked-salmon-and-farfalle-in-lemon-cream-sauce-321798">Smoked Salmon and Farfalle in Lemon Cream Sauce</a>.',
    },
    {
      id: 657537,
      title: 'Quick and Easy Caprese Salad',
      image: 'https://img.spoonacular.com/recipes/657537-556x370.jpg',
      summary:
        'Need a <b>gluten free, lacto ovo vegetarian, and primal main course</b>? Quick and Easy Caprese Salad could be an awesome recipe to try. One serving contains <b>350 calories</b>, <b>15g of protein</b>, and <b>27g of fat</b>. For <b>$1.97 per serving</b>, this recipe <b>covers 20%</b> of your daily requirements of vitamins and minerals. This recipe serves 4. It is brought to you by Foodista. A mixture of balsamic vinegar, basil leaves, bell pepper, and a handful of other ingredients are all it takes to make this recipe so delicious. 3 people found this recipe to be yummy and satisfying. It is an <b>affordable</b> recipe for fans of Mediterranean food. From preparation to the plate, this recipe takes approximately <b>45 minutes</b>. Overall, this recipe earns a <b>good spoonacular score of 74%</b>. Try <a href="https://spoonacular.com/recipes/quick-caprese-salad-551296">Quick Caprese Salad</a>, <a href="https://spoonacular.com/recipes/quick-roasted-tomato-caprese-pasta-salad-673329">Quick Roasted Tomato Caprese Pasta Salad</a>, and <a href="https://spoonacular.com/recipes/easy-caprese-salad-524331">Easy Caprese Salad</a> for similar recipes.',
    },
  ],
};
