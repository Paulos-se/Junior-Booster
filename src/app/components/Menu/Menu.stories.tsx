import type { Meta, StoryObj } from '@storybook/react';
import { FadeMenu as Menu, MenuProps } from './Menu'; // Assuming `FadeMenu` is aliased as `Menu`
import { menuItemsMock } from '../../testData/menu';

interface Args {
  menuItems: MenuProps['menuItems'];
  color: string;
}

const meta: Meta<typeof Menu> = {
  title: 'Menu',
  component: Menu,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<Args>;

export const Primary: Story = {
  args: {
    menuItems: menuItemsMock.menuItems,
    color: menuItemsMock.color,
  },
  
};
