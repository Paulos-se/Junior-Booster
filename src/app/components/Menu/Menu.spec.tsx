import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { FadeMenu as Menu } from './Menu'; // Adjust path as necessary
import { menuItemsMock } from '../../testData/menu'; // Adjust path as necessary
import { screen } from '@testing-library/dom';

describe('Menu', () => {
  const { menuItems, color } = menuItemsMock;
  it('renders correctly', () => {
    const { baseElement } = render(
      <Menu menuItems={menuItems} color={color} />
    );
    expect(baseElement).toBeTruthy();
  });
  it('matches snapshot', () => {
    render(<Menu menuItems={menuItems} color={color} />);
    const button = screen.getByText('Button');
    fireEvent.click(button);
    expect(document.body).toMatchSnapshot();
  });
});
