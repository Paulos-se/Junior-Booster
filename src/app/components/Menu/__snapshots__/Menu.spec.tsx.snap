// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches snapshot 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      aria-controls="fade-menu"
      aria-expanded="true"
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      id="fade-button"
      tabindex="0"
      type="button"
    >
      Button
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    id="fade-menu"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper MuiMenu-paper MuiMenu-paper css-3dzjca-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        aria-labelledby="fade-button"
        class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-yvn3nr-MuiTypography-root-MuiLink-root"
            href="/"
          >
            <div
              class="MuiBox-root css-17s5tg5"
            >
              <div
                class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="HomeIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                  />
                </svg>
              </div>
              Home
            </div>
          </a>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-yvn3nr-MuiTypography-root-MuiLink-root"
            href="/about"
          >
            <div
              class="MuiBox-root css-17s5tg5"
            >
              <div
                class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="InfoIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
                  />
                </svg>
              </div>
              About
            </div>
          </a>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-yvn3nr-MuiTypography-root-MuiLink-root"
            href="/contact"
          >
            <div
              class="MuiBox-root css-17s5tg5"
            >
              <div
                class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ContactMailIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21 8V7l-3 2-3-2v1l3 2zm1-5H2C.9 3 0 3.9 0 5v14c0 1.1.9 2 2 2h20c1.1 0 1.99-.9 1.99-2L24 5c0-1.1-.9-2-2-2M8 6c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3m6 12H2v-1c0-2 4-3.1 6-3.1s6 1.1 6 3.1zm8-6h-8V6h8z"
                  />
                </svg>
              </div>
              Contact
            </div>
          </a>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
